### PR TITLE
io/npz.py improved functionality

### PR DESCRIPTION
**Pull request by SeRealMF**
# Overview
Previously, _npz.py_ had limited possibilities, especially compared to other input/output types like _json.py_. 
The goal of the update was to improve the functionality of _npz.py_, as visible below.

## Addressed issues
This update is referring to the following issue [here](https://github.com/tataratat/splinepy/issues/262).

The added functionalities are:
-  allowing bezier types
-  returning initialized spline
-  casting knot_vectors to list
-  round trip test
- allowing multiple splines

## Checklists
* [ ] Documentations are up-to-date.
* [ ] Added example(s)
* [x] Added test(s)

### CODE REVIEW

File: splinepy/io/npz.py [hunk 1]
```diff
@@ -10,6 +10,8 @@
 
 import numpy as np
 
+import splinepy as sp
```
Comment (j042) [inline splinepy/io/npz.py:13]: please don't import the whole library. instead, import the modules you need

File: splinepy/io/npz.py [hunk 2]
```diff
@@ -23,22 +25,27 @@ def load(
...
-    dict_spline: dict
+    list_spline: list
     """
     loaded = np.load(fname)
-    whatami = loaded["whatami"][0]
 
     dict_spline = {}
-    # update weights
-    if whatami.startswith("NURBS"):
+    dict_spline.update(
```
Comment (j042) [inline splinepy/io/npz.py:33]: can you change all `update()` to `__getitem__`? (`dict_spline["key"] = value`)

File: tests/test_npz.py [hunk 3]
```diff
@@ -0,0 +1,57 @@
+import tempfile
+
+try:
+    from . import common as c
+except BaseException:
+    import common as c
+
+
+class npzExportTest(c.unittest.TestCase):
+    def test_gismo_import(self):
```
Comment (j042) [inline tests/test_npz.py:10]: this isn't a gismo import test

**Review by j042 — CHANGES_REQUESTED** [review]

File: splinepy/io/npz.py [hunk 4]
```diff
@@ -10,60 +10,108 @@
...
-    whatami = loaded["whatami"][0]
 
-    dict_spline = {}
-    # update weights
-    if whatami.startswith("NURBS"):
-        dict_spline.update(weights=loaded["weights"])
+    # Initialize an empty list to store the splines
+    list_of_spline_dicts = []
+
+    processed_keys = set()
```
Comment (j042) [inline splinepy/io/npz.py:35]:
```suggestion
```
unused

File: splinepy/io/npz.py [hunk 5]
```diff
@@ -10,60 +10,108 @@
...
+            # Append dictionary of spline to the list
+            list_of_spline_dicts.append(dict_spline)
 
+            # Add keys to processed_keys set
+            processed_keys.add(f"spline_{i}_control_points")
+            processed_keys.add(f"spline_{i}_degrees")
+            processed_keys.add(f"spline_{i}_weights")
+            processed_keys.add(f"spline_{i}_whatami")
+            for j in range(dict_spline["degrees"].size):
+                processed_keys.add(f"spline_{i}_knot_vectors_{j}")
```
Comment (j042) [inline splinepy/io/npz.py:73]:
```suggestion
```
unused

File: splinepy/io/npz.py [hunk 6]
```diff
@@ -10,60 +10,108 @@
...
+
+        # Add the weights if the spline is rational
+        if spline.is_rational:
+            property_dicts[f"{prefix}weights"] = spline.weights
 
-    if spline.whatami.startswith("NURBS"):
-        property_dicts.update(weights=spline.weights)
+        # Add the knot vectors if the spline has them
+        if spline.has_knot_vectors:
+            for j in range(spline.degrees.size):
```
Comment (j042) [inline splinepy/io/npz.py:109]:
```suggestion
            for j in range(spline.para_dim):
```

File: splinepy/io/npz.py [hunk 7]
```diff
@@ -10,60 +10,108 @@
...
     -----------
     fname: str
-    spline: BSpline or NURBS
+    splines: list of NURBS / BSpline / Bezier / Rational Bezier
 
     Returns
     --------
     None
     """
-    property_dicts = {
```
Comment (j042) [inline splinepy/io/npz.py:57]: can you add a check here, if `list_of_splines` is a list? if it is a single spline, you can create a list. Else error.

File: tests/test_npz.py [hunk 8]
```diff
@@ -0,0 +1,57 @@
...
+        nur_el3.elevate_degrees(1)
+        bsp_el2.insert_knots(1, [0.5])
+        nur_el3.insert_knots(1, [0.5])
+
+        # Test output against input
+        list_of_splines = [bez_el0, rbz_el1, bsp_el2, nur_el3]
+        with tempfile.TemporaryDirectory() as tmpd:
+            tmpf = c.to_tmpf(tmpd)
+            c.splinepy.io.npz.export(tmpf, list_of_splines)
+            list_of_splines_loaded = c.splinepy.io.npz.load(tmpf + ".npz")
```
Comment (j042) [inline tests/test_npz.py:47]: can you check if this is an expected behavior of `np.savez`, that they always automatically append `.npz` if no extension is specified?

File: splinepy/io/npz.py [hunk 9]
```diff
@@ -10,60 +10,108 @@
...
+
+    # Loop through the list of splines and add their properties to the dictionary
+    for i, spline in enumerate(list_of_splines):
+        # Use a prefix to distinguish different splines
+        prefix = f"spline_{i}_"
+
+        # Add the common properties of all splines
+        property_dicts[f"{prefix}degrees"] = spline.degrees
+        property_dicts[f"{prefix}control_points"] = spline.control_points
+        property_dicts[f"{prefix}whatami"] = np.array([spline.whatami])
```
Comment (j042) [inline splinepy/io/npz.py:101]:
```suggestion
```
we don't use this information for now

**Review by j042 — CHANGES_REQUESTED** [review]

File: splinepy/io/npz.py (hunk 4 above)
Comment (SeRealMF) [reply splinepy/io/npz.py:35]: The processed_keys are used in
```
for key in loaded:
    if key.startswith("spline_") and key not in processed_keys:
```

File: splinepy/io/npz.py (hunk 5 above)
Comment (SeRealMF) [reply splinepy/io/npz.py]: The processed_keys are used in
```
for key in loaded:
    if key.startswith("spline_") and key not in processed_keys:
```

File: tests/test_npz.py (hunk 8 above)
Comment (SeRealMF) [reply tests/test_npz.py:47]: Fixed in [866438a](https://github.com/tataratat/splinepy/pull/280/commits/866438ab03a10e4e3be03aa1f919e4b0c83b42c0)

File: splinepy/io/npz.py (hunk 9 above)
Comment (SeRealMF) [reply splinepy/io/npz.py]: Fixed in [08105e1](https://github.com/tataratat/splinepy/pull/280/commits/08105e1163ee942d3dc2d9e6c02498f0fa925718)

File: splinepy/io/npz.py (hunk 7 above)
Comment (SeRealMF) [reply splinepy/io/npz.py:57]: Fixed in [866438a](https://github.com/tataratat/splinepy/pull/280/commits/866438ab03a10e4e3be03aa1f919e4b0c83b42c0)

File: splinepy/io/npz.py (hunk 4 above)
Comment (j042) [reply splinepy/io/npz.py:35]: I see. Instead, can you just export number of spline and and loop range(n_splines)? This seems a bit much

Comment (SeRealMF) [reply splinepy/io/npz.py]: Should be fixed in [e8d3693](https://github.com/tataratat/splinepy/pull/280/commits/e8d3693c8a70f2d15d01e19017b1f7fb4f990877)

File: splinepy/io/npz.py [hunk 10]
```diff
@@ -10,60 +10,102 @@
...
+        if isinstance(list_of_splines, splinepy_core.PySpline):
+            list_of_splines = [list_of_splines]
+        else:
+            raise TypeError("Only a list of splines or a spline can be saved.")
+
+    # Initialize an empty dictionary to store the properties of each spline
+    property_dicts = {}
+
+    # Add number of splines in list to dict
+    property_dicts["num_of_splines"] = len(list_of_splines)
```
Comment (j042) [inline splinepy/io/npz.py:86]:
```suggestion
    property_dicts["number_of_splines"] = len(list_of_splines)
```

File: splinepy/io/npz.py [hunk 11]
```diff
@@ -10,60 +10,102 @@
...
 
-    dict_spline = {}
-    # update weights
-    if whatami.startswith("NURBS"):
-        dict_spline.update(weights=loaded["weights"])
+    # Initialize an empty list to store the splines
+    list_of_spline_dicts = []
+
+    # Loop through keys in loaded dictionary and find the ones starting with 'spline_'
+    for i in range(loaded["num_of_splines"]):
```
Comment (j042) [inline splinepy/io/npz.py:36]:
```suggestion
    for i in range(loaded["number_of_splines"]):
```

File: splinepy/io/npz.py (hunk 4 above)
Comment (j042) [reply splinepy/io/npz.py]: thanks!

**Review by j042 — APPROVED** [review]
Thanks!

**Pull request merged**